src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceSetDeserializer.cs

```diff
@@ -124,10 +126,133 @@ public virtual IEnumerable ReadResourceSet(ODataResourceSetWrapper resourceSet,
                     Error.Format(SRResources.TypeCannotBeDeserialized, elementType.FullName()));
             }
 
-            foreach (ODataResourceWrapper resourceWrapper in resourceSet.Resources)
+            Collection<object> coll = new Collection<object>();
```

Comment:
Any reason you changed from yield return to creating the entire collection before returning?